assembly/runtime.ts

```diff
@@ -845,125 +895,153 @@ export class RootContext extends BaseContext {
   onResponseMetadata(a: u32): FilterMetadataStatusValues { return FilterMetadataStatusValues.Continue }
-  onResponseBody(body_buffer_length: size_t, end_of_stream: bool): FilterDataStatusValues { return FilterDataStatusValues.Continue }
+  onResponseBody(body_buffer_length: usize, end_of_stream: bool): FilterDataStatusValues { return FilterDataStatusValues.Continue }
   onResponseTrailers(s: u32): FilterTrailersStatusValues { return FilterTrailersStatusValues.Continue }
-  onLog(): void { }  // Called after onDone when logging is requested.
+  
+  // Called after onDone when logging is requested.
+  onLog(): void { 
+    log(LogLevelValues.debug, "context id: " + this.context_id.toString() + ": onLog()");
+  }
 }
 
 function get_plugin_root_id(): string {
-
   let root_id = get_property("plugin_root_id");
   if (root_id.byteLength == 0) {
     return "";
   }
   return String.UTF8.decode(root_id);
 }
 
+let root_context_factory_map = new Map<string, (context_id: u32) => RootContext>();
+
 let context_map = new Map<u32, BaseContext>();
+
+//create root context if doesn't exist
 export function ensureRootContext(root_context_id: u32): RootContext {
+  log(LogLevelValues.debug, "ensureRootContext(root_context_id: " + root_context_id.toString() + ")");
+  log(LogLevelValues.debug, "Current context_map: " + context_map.keys().join(", "));
   if (context_map.has(root_context_id)) {
+    log(LogLevelValues.debug, "Returning root context for id: " + root_context_id.toString());
     return getRootContext(root_context_id);
   }
-  let root_id = get_plugin_root_id();
-  if (root_factory.has(root_id)) {
-    let root_context_func = root_factory.get(root_id);
-    let root_context = root_context_func();
-    root_context.context_id = root_context_id;
-    context_map.set(root_context_id, root_context);
-
-    log(LogLevelValues.warn, "returning context for " + root_id);
-    return root_context;
+  const root_id = get_plugin_root_id();
+  log(LogLevelValues.debug, "Registering new root context for " + root_id + " with id: " + root_context_id.toString());
+  if (!root_context_factory_map.has(root_id)) {
+    throw new Error("Missing root context factory for root id: " + root_id);
   }
-
-  log(LogLevelValues.warn, "did not find root id " + root_id)
-
-  let root_context = new RootContext();
+  const root_context_func = root_context_factory_map.get(root_id);
+  const root_context = root_context_func(root_context_id);
   root_context.context_id = root_context_id;
   context_map.set(root_context_id, root_context);
   return root_context;
 }
 
-let root_factory = new Map<string, () => RootContext>();
-//let context_map = new Map<u32, Context>();
+// create a context if doesnt exist.
+export function ensureContext(context_id: u32, root_context_id: u32): void {
+  log(LogLevelValues.debug, "ensureContext(context_id: " + context_id.toString() + ", root_context_id: " + root_context_id.toString() + ")");
```

Comment:
Ensure context and ensureRootContext wasn't working properly before. It was better now with the context_ids being passed down to the factories.